### PR TITLE
PHPStan fixes.

### PR DESCRIPTION
**Pull request by dereuromark**
In my experience PHPStan level 5 is a good target to reach, as it finds important interface contracting issues, while not being too pedantic about nullable etc.
But we are still quite a bit away from reaching this clean code state yet.

I managed to get level 5 fails down from 54 to 45, remaining is now:

```
 ------ ------------------------------------------------------------------------------------------ 
  Line   Collection/CollectionTrait.php (in context of class Cake\Collection\Collection)           
 ------ ------------------------------------------------------------------------------------------ 
  528    Parameter #1 $it of method AppendIterator::append() expects Iterator, Traversable given.  
  529    Parameter #1 $it of method AppendIterator::append() expects Iterator, Traversable given.  
 ------ ------------------------------------------------------------------------------------------ 

 ------ -------------------------------------------------------------------------------------------- 
  Line   Collection/CollectionTrait.php (in context of class Cake\Collection\Iterator\TreeIterator)  
 ------ -------------------------------------------------------------------------------------------- 
  528    Parameter #1 $it of method AppendIterator::append() expects Iterator, Traversable given.    
  529    Parameter #1 $it of method AppendIterator::append() expects Iterator, Traversable given.    
 ------ -------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------- 
  Line   Collection/CollectionTrait.php (in context of class Cake\Collection\Iterator\TreePrinter)  
 ------ ------------------------------------------------------------------------------------------- 
  528    Parameter #1 $it of method AppendIterator::append() expects Iterator, Traversable given.   
  529    Parameter #1 $it of method AppendIterator::append() expects Iterator, Traversable given.   
 ------ ------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------- 
  Line   Collection/CollectionTrait.php (in context of class Cake\Collection\Iterator\ZipIterator)  
 ------ ------------------------------------------------------------------------------------------- 
  528    Parameter #1 $it of method AppendIterator::append() expects Iterator, Traversable given.   
  529    Parameter #1 $it of method AppendIterator::append() expects Iterator, Traversable given.   
 ------ ------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------ 
  Line   Collection/CollectionTrait.php (in context of class Cake\ORM\ResultSet)                   
 ------ ------------------------------------------------------------------------------------------ 
  528    Parameter #1 $it of method AppendIterator::append() expects Iterator, Traversable given.  
  529    Parameter #1 $it of method AppendIterator::append() expects Iterator, Traversable given.  
 ------ ------------------------------------------------------------------------------------------ 

 ------ ---------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   Database/TypeConverterTrait.php (in context of class Cake\Database\Statement\BufferedStatement)                                               
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------- 
  38     Parameter #2 $driver of method Cake\Database\TypeInterface::toDatabase() expects Cake\Database\Driver, Cake\Database\DriverInterface given.   
  39     Parameter #2 $driver of method Cake\Database\TypeInterface::toStatement() expects Cake\Database\Driver, Cake\Database\DriverInterface given.  
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------------- 
  Line   Filesystem/Filesystem.php                                                                          
 ------ --------------------------------------------------------------------------------------------------- 
  115    Parameter #1 $it of class RegexIterator constructor expects Iterator, Traversable given.           
  118    Parameter #1 $it of class CallbackFilterIterator constructor expects Iterator, Traversable given.  
 ------ --------------------------------------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   Http/ActionDispatcher.php                                                                                                                     
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------- 
  62     Parameter #2 $response of method Cake\Http\ControllerFactory::create() expects Cake\Http\Response, Psr\Http\Message\ResponseInterface given.  
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   Http/BaseApplication.php                                                                                                                                
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------- 
  242    Parameter #1 $request of method Cake\Http\ActionDispatcher::dispatch() expects Cake\Http\ServerRequest, Psr\Http\Message\ServerRequestInterface given.  
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   Http/Middleware/CsrfProtectionMiddleware.php                                                                                                                                                  
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  106    Parameter #2 $request of method Cake\Http\Middleware\CsrfProtectionMiddleware::_addTokenToRequest() expects Cake\Http\ServerRequest, Psr\Http\Message\ServerRequestInterface given.           
  109    Parameter #3 $response of method Cake\Http\Middleware\CsrfProtectionMiddleware::_addTokenCookie() expects Cake\Http\Response, Psr\Http\Message\ResponseInterface given.                       
  111    Parameter #1 $request of method Cake\Http\Middleware\CsrfProtectionMiddleware::_validateAndUnsetTokenField() expects Cake\Http\ServerRequest, Psr\Http\Message\ServerRequestInterface given.  
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ -------------------------------------------------------------------------- 
  Line   I18n/Parser/MoFileParser.php                                              
 ------ -------------------------------------------------------------------------- 
  63     Parameter #1 $filename of function fopen expects string, resource given.  
 ------ -------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   ORM/EagerLoader.php                                                                                                                                                
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  620    Parameter #1 $repository of method Cake\ORM\EagerLoader::externalAssociations() expects Cake\ORM\Table, Cake\Datasource\RepositoryInterface given.                 
  626    Parameter #3 $statement of method Cake\ORM\EagerLoader::_collectKeys() expects Cake\Database\Statement\BufferedStatement, Cake\Database\StatementInterface given.  
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  Line   ORM/Marshaller.php                                                                                                                                                                                                
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 
  320    Parameter #1 $assoc of method Cake\ORM\Marshaller::_belongsToMany() expects Cake\ORM\Association\BelongsToMany, Cake\ORM\Association given.                                                                       
  741    Parameter #1 $entities of method Cake\ORM\Marshaller::mergeMany() expects iterable<Cake\Datasource\EntityInterface>&Traversable, (array<Cake\Datasource\EntityInterface>)|Cake\Datasource\EntityInterface given.  
  769    Parameter #1 $entities of method Cake\ORM\Marshaller::mergeMany() expects iterable<Cake\Datasource\EntityInterface>&Traversable, array<Cake\Datasource\EntityInterface> given.                                    
  772    Parameter #2 $assoc of method Cake\ORM\Marshaller::_mergeJoinData() expects Cake\ORM\Association\BelongsToMany, Cake\ORM\Association given.                                                                       
  808    Parameter #1 $entities of method Cake\ORM\Marshaller::mergeMany() expects iterable<Cake\Datasource\EntityInterface>&Traversable, array<Cake\Datasource\EntityInterface> given.                                    
 ------ ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ 

 ------ ---------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   ORM/Query.php                                                                                                                                 
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------- 
  174    Parameter #1 $table of method Cake\ORM\Query::addDefaultTypes() expects Cake\ORM\Table, Cake\Datasource\RepositoryInterface given.            
  445    Parameter #1 $table of method Cake\ORM\Query::_addAssociationsToTypeMap() expects Cake\ORM\Table, Cake\Datasource\RepositoryInterface given.  
  556    Parameter #1 $table of method Cake\ORM\Query::_addAssociationsToTypeMap() expects Cake\ORM\Table, Cake\Datasource\RepositoryInterface given.  
  633    Parameter #1 $table of method Cake\ORM\Query::_addAssociationsToTypeMap() expects Cake\ORM\Table, Cake\Datasource\RepositoryInterface given.  
  682    Parameter #1 $table of method Cake\ORM\Query::_addAssociationsToTypeMap() expects Cake\ORM\Table, Cake\Datasource\RepositoryInterface given.  
  747    Parameter #1 $table of method Cake\ORM\Query::_addAssociationsToTypeMap() expects Cake\ORM\Table, Cake\Datasource\RepositoryInterface given.  
 ------ ---------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------------------ 
  Line   ORM/ResultSet.php                                                                                                                         
 ------ ------------------------------------------------------------------------------------------------------------------------------------------ 
  378    Parameter #1 $table of method Cake\ORM\EagerLoader::associationsMap() expects Cake\ORM\Table, Cake\Datasource\RepositoryInterface given.  
 ------ ------------------------------------------------------------------------------------------------------------------------------------------ 

 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   ORM/Table.php                                                                                                                                                  
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  490    Parameter #1 $schema of method Cake\ORM\Table::_initializeSchema() expects Cake\Database\Schema\TableSchema, Cake\Database\Schema\TableSchemaInterface given.  
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   Routing/Router.php                                                                                                                                         
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------- 
  241    Parameter #1 $request of static method Cake\Routing\Router::pushRequest() expects Cake\Http\ServerRequest, Psr\Http\Message\ServerRequestInterface given.  
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   TestSuite/ConsoleIntegrationTestTrait.php (in context of class Cake\TestSuite\ConsoleIntegrationTestCase)                                            
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------- 
  99     Parameter #3 $in of class Cake\Console\ConsoleIo constructor expects Cake\Console\ConsoleInput|null, PHPUnit\Framework\MockObject\MockObject given.  
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   TestSuite/Fixture/TestFixture.php                                                                                                                                          
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  280    Parameter #1 $connection of method Cake\Database\Schema\SqlGeneratorInterface::createSql() expects Cake\Database\Connection, Cake\Datasource\ConnectionInterface given.    
  315    Parameter #1 $connection of method Cake\Database\Schema\SqlGeneratorInterface::dropSql() expects Cake\Database\Connection, Cake\Datasource\ConnectionInterface given.      
  430    Parameter #1 $connection of method Cake\Database\Schema\SqlGeneratorInterface::truncateSql() expects Cake\Database\Connection, Cake\Datasource\ConnectionInterface given.  
 ------ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

```

### CODE REVIEW

File: src/Utility/Xml.php [hunk 1]
```diff
@@ -277,6 +277,7 @@ public static function fromArray($input, array $options = [])
         ];
         $options += $defaults;
 
+        /** @var DOMDocument|\DOMElement $dom */
```
Comment (ADmad) [inline src/Utility/Xml.php]: Backslash missing before `DOMDocument`. It's strange the var needs to be hinted this way when it's obviously a `DOMDocument` instance.

Comment (dereuromark) [reply src/Utility/Xml.php]: The whole DOM and in general PHP Stuff is really badly written, lots of legacy I am sure.

Comment (ADmad) [reply src/Utility/Xml.php]: Yeah the DOM classes are really wacky.

**Comment by ADmad** [comment]
Some of these changes make psalm unhappy.

**Comment by ADmad** [comment]
psalm is now happy and the errors shouldn't have increased for phpstan.

**Pull request merged**